File: CarHomeWidget/src/main/java/com/anod/car/home/backup/RestoreTask.java
package com.anod.car.home.backup;

import android.net.Uri;
import android.os.AsyncTask;

/**
 * @author alex
 * @date 12/30/13
 */
public class RestoreTask extends AsyncTask<Uri, Void, Integer> {
    public static final String SCHEME_FILE = "file";
    private int mType;
    private PreferencesBackupManager mBackupManager;
    private int mAppWidgetId;
    private RestoreTaskListener mListener;

    public interface RestoreTaskListener {
        void onRestorePreExecute(int type);
        void onRestoreFinish(int type, int code);
    }

    public RestoreTask(int type, PreferencesBackupManager backupManager, int appWidgetId,
            RestoreTaskListener listener) {
        mType = type;
        mBackupManager = backupManager;
        mAppWidgetId = appWidgetId;
        mListener = listener;
    }

    @Override
    protected void onPreExecute() {
        mListener.onRestorePreExecute(mType);
    }

    protected Integer doInBackground(Uri... uris) {
        Uri uri = uris[0];
        if (mType == PreferencesBackupManager.TYPE_INCAR) {

            if (SCHEME_FILE.equals(uri.getScheme())) {
                return mBackupManager.doRestoreInCarLocal(uri.getPath());
            }

            return mBackupManager.doRestoreInCarUri(uri);
        }

        if (SCHEME_FILE.equals(uri.getScheme())) {
            return mBackupManager.doRestoreWidgetLocal(uri.getPath(), mAppWidgetId);
        }
        return mBackupManager.doRestoreWidgetUri(uri, mAppWidgetId);
    }

    @Override
    protected void onPostExecute(Integer result) {
        mListener.onRestoreFinish(mType, result);
    }

}

File: CarHomeWidget/src/main/java/com/anod/car/home/backup/RestoreTask.kt
package com.anod.car.home.backup

import android.content.ContentResolver
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * @author alex
 * @date 12/30/13
 */
class RestoreTask(
        private val type: Int,
        private val backupManager: PreferencesBackupManager,
        private val appWidgetId: Int,
        private val uri: Uri) {

    constructor(backupManager: PreferencesBackupManager, appWidgetId: Int, uri: Uri)
            : this(PreferencesBackupManager.TYPE_MAIN, backupManager, appWidgetId, uri)

    constructor(backupManager: PreferencesBackupManager, uri: Uri)
            : this(PreferencesBackupManager.TYPE_INCAR, backupManager, 0, uri)

    suspend fun execute(): Int = withContext(Dispatchers.IO) {
        if (type == PreferencesBackupManager.TYPE_INCAR) {
            return@withContext if (ContentResolver.SCHEME_FILE == uri.scheme) {
                backupManager.doRestoreInCarLocal(uri)
            } else backupManager.doRestoreInCarUri(uri)
        }

        return@withContext if (ContentResolver.SCHEME_FILE == uri.scheme) {
            backupManager.doRestoreWidgetLocal(uri, appWidgetId)
        } else backupManager.doRestoreWidgetUri(uri, appWidgetId)
    }

}